From 9262f919da754c2639b721e278da5daabdaada11 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Maxime=20M=C3=A9r=C3=A9?= <maxime.mere@foss.st.com>
Date: Fri, 12 Sep 2025 15:50:51 +0200
Subject: [PATCH 908/954] fix(stm32mp2): handle power down on secondary core

Call psci_pwrdown_cpu for each core to correctly handle coherency
and allow to stop the cortex A35 cluster, including the L2 cache,
and to allow Standby2 entry.

On a PSCI shutdown request, only one core gets an opportunity to
execute the powerdown sequence, with this path this CPU raises a
power down SGI to other online cores to handle the required
operations to cleanly deactivate the system (deactivate the SMP
for example).

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/arm-trusted-firmware.git 60ed10cb0ccf349a4f19c7a3fe5bdc352191dcdc fix(stm32mp2): handle power down on secondary core]
Change-Id: I1d978e493749d2417ee837e33ae8335ddd80af11
Signed-off-by: Patrick Delaunay <patrick.delaunay@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/tf-a/+/473802
Reviewed-by: Maxime MERE <maxime.mere@foss.st.com>
Tested-by: Maxime MERE <maxime.mere@foss.st.com>
Domain-Review: Maxime MERE <maxime.mere@foss.st.com>
---
 plat/st/stm32mp2/stm32mp2_pm.c | 115 ++++++++++++++++++++++++++++-----
 1 file changed, 98 insertions(+), 17 deletions(-)

diff --git a/plat/st/stm32mp2/stm32mp2_pm.c b/plat/st/stm32mp2/stm32mp2_pm.c
index 2ae7a2cc50..e4dbbdc5d2 100644
--- a/plat/st/stm32mp2/stm32mp2_pm.c
+++ b/plat/st/stm32mp2/stm32mp2_pm.c
@@ -34,6 +34,10 @@
 #include "../../../lib/psci/psci_private.h"
 #include "scmi_private.h"
 
+#ifndef STM32MP_PWRDOWN_SGI
+#define STM32MP_PWRDOWN_SGI ARM_IRQ_SEC_SGI_7
+#endif
+
 /* Default value for STM32MP25 with STPMIC25, defined in AN5727 */
 #define DEFAULT_POPL_D1		3U
 #define DEFAULT_PODH_D2		1U
@@ -894,6 +898,80 @@ static void __dead2 stm32_pwr_domain_pwr_down_wfi(const psci_power_state_t
 	panic();
 }
 
+#if !STM32MP21
+/*
+ * For a shutdown, each CA35 core in the system should do their power down
+ * sequence.
+ * On a PSCI shutdown/reboot request, only one core gets an
+ * opportunity to do the powerdown sequence, to achieve Standby entry sequence,
+ * this CPU raise a power down SGI to rest of the core  which are online and
+ * add in its handler managed the core power down sequence (remove core in
+ * coherency) to allow WFI for the cluster, including the L2 cache.
+ */
+static uint64_t stm32_pwrdwn_sgi_handler(uint32_t id, uint32_t flags, void *handle, void *cookie)
+{
+	u_register_t mpidr = read_mpidr();
+	unsigned int core_id = MPIDR_AFFLVL0_VAL(mpidr);
+
+	INFO("Power down core %u\n", core_id);
+
+	/* Deactivate SGI */
+	plat_ic_end_of_interrupt(STM32MP_PWRDOWN_SGI);
+
+	/* Disable GIC CPU interface to prevent waking up from WFI. */
+	stm32mp_gic_cpuif_disable();
+
+	/* Prepare power off, set SMPEN=0 on this core to allow Standby */
+	psci_pwrdown_cpu(PLAT_MAX_PWR_LVL);
+
+	/* Core is no more running */
+	stm32mp_state_set(STM32MP_SECONDARY_CPU, STATE_RUNNING, false);
+	sev();
+
+	dmbsy();
+
+	psci_power_down_wfi();
+
+	/* This shouldn't be reached */
+	panic();
+
+	return 0;
+}
+
+/*
+ * Setup the SGI interrupt that will be used trigger the execution of power
+ * down sequence for all the secondary cores. This interrupt is setup to be
+ * handled in EL3 context at a priority defined by the platform.
+ */
+static void stm32_setup_cpu_pwrdown_sgi(void)
+{
+	u_register_t flags;
+	int ret = 0;
+
+	plat_ic_set_interrupt_type(STM32MP_PWRDOWN_SGI, INTR_TYPE_S_EL1);
+	plat_ic_set_interrupt_priority(STM32MP_PWRDOWN_SGI, GIC_HIGHEST_SEC_PRIORITY);
+
+	flags = 0;
+	set_interrupt_rm_flag(flags, NON_SECURE);
+	ret = register_interrupt_type_handler(INTR_TYPE_S_EL1, stm32_pwrdwn_sgi_handler, flags);
+	if (ret != 0) {
+		ERROR("Register interrupt failed %d\n", ret);
+		panic();
+	}
+
+	plat_ic_enable_interrupt(STM32MP_PWRDOWN_SGI);
+}
+
+/*
+ * Callback function to raise a SGI designated to trigger the CPU power down
+ * sequence on the secondary core.
+ */
+static void stm32_raise_pwrdown_sgi(u_register_t mpidr)
+{
+	plat_ic_raise_s_el1_sgi(STM32MP_PWRDOWN_SGI, mpidr);
+}
+#endif
+
 static void __dead2 stm32_system_off(void)
 {
 	uintptr_t pwr_base = stm32mp_pwr_base();
@@ -914,22 +992,21 @@ static void __dead2 stm32_system_off(void)
 
 #if !STM32MP21
 	if (!stm32_freeze_other_core(core_id)) {
-		VERBOSE("PSCI system off with other core running.\n");
-
-		/* Core is no more running */
-		stm32mp_state_set(STM32MP_SECONDARY_CPU, STATE_RUNNING, false);
-
-		/* Program secondary CPU entry points */
-		stm32mp_ca35_set_vbar(stm32_sec_entrypoint);
-
-		/* After reset, the core is stopped, waiting in WFI loop */
-		stm32mp_state_set(STM32MP_SECONDARY_CPU, STATE_START, false);
+		int ret;
 
-		/* Reset the secondary core */
-		mmio_write_32(RCC_BASE + RCC_C1P1RSTCSETR, RCC_C1P1RSTCSETR_C1P1PORRST);
-
-		/* Forbid access to DDR */
+		VERBOSE("PSCI system off with other core running.\n");
+		/* Forbid access on DDR in stm32_cpu_standby() */
 		stm32mp_state_set(STM32MP_PRIMARY_CPU, STATE_DDR, false);
+		/* Send powerdown request to online secondary core(s) */
+		stm32_setup_cpu_pwrdown_sgi();
+		ret = psci_stop_other_cores(0, stm32_raise_pwrdown_sgi);
+		if (ret != PSCI_E_SUCCESS) {
+			ERROR("Failed to powerdown the secondary core\n");
+			panic();
+		}
+		while (stm32mp_state_check(STM32MP_SECONDARY_CPU, STATE_RUNNING)) {
+			wfe();
+		}
 	}
 #endif /* !STM32MP21 */
 
@@ -965,7 +1042,7 @@ static void __dead2 stm32_system_off(void)
 		}
 	}
 #endif
-
+	/* Flush and invalidate data cache before DDR deactivation */
 	dcsw_op_all(DCCISW);
 
 #if !STM32MP_M33_TDCID
@@ -1000,7 +1077,6 @@ static void __dead2 stm32_system_off(void)
 	/* Do not maintain RETRAM memory content in Standby or Vbat */
 	mmio_write_32(pwr_base + PWR_CR10, PWR_CR10_RETRBSEN_DISABLE);
 #endif
-
 	/* Clear PM context in BKPSRAM: cold boot at next wake-up */
 	stm32_pm_context_clear();
 
@@ -1033,10 +1109,15 @@ static void __dead2 stm32_system_off(void)
 	stm32mp_state_set(core_id, STATE_RUNNING, false);
 
 	INFO("BL31: power off\n");
+	console_flush();
+
+	stm32mp_ca35_lpi_isolate();
+
+	psci_pwrdown_cpu(PLAT_MAX_PWR_LVL);
 
 	dsb();
 	isb();
-	wfi();
+	psci_power_down_wfi();
 
 	/* This shouldn't be reached */
 	panic();
-- 
2.53.0

